packages/flutter_tools/lib/src/android/android_studio.dart

```diff
@@ -0,0 +1,257 @@
+
+    bool _hasStudioAt(String path, {String newerThan}) {
+      return studios.any((AndroidStudio studio) {
+        if (studio.directory != path) return false;
+        if (newerThan != null) {
+          return studio.version.compareTo(newerThan) >= 0;
+        }
+        return true;
+      });
+    }
+
+    // Read all $HOME/AndroidStudio*/system/.home files. There may be several
+    // pointing to the same installation, so we grab only the latest one.
+    for (FileSystemEntity entity in fs.directory(homeDirPath).listSync()) {
+      if (entity is Directory && entity.basename.startsWith('.AndroidStudio')) {
+        AndroidStudio studio = new AndroidStudio.fromHomeDot(entity);
+        if (studio != null &&
+            !_hasStudioAt(studio.directory, newerThan: studio.version)) {
+          studios.removeWhere(
+              (AndroidStudio other) => other.directory == studio.directory);
+          studios.add(studio);
+        }
+      }
+    }
+
+    String configuredStudioDir = config.getValue('android-studio-dir');
+    if (configuredStudioDir != null && !_hasStudioAt(configuredStudioDir)) {
+      studios.add(new AndroidStudio(configuredStudioDir,
+          configured: configuredStudioDir));
+    }
+
+    if (platform.isLinux) {
+      void _checkWellKnownPath(String path) {
+        if (fs.isDirectorySync(path) && !_hasStudioAt(path)) {
+          studios.add(new AndroidStudio(path));
+        }
+      }
+
+      // Add /opt/android-studio and $HOME/android-studio, if they exist.
+      _checkWellKnownPath('/opt/android-studio');
+      _checkWellKnownPath('$homeDirPath/android-studio');
+    }
+    return studios;
+  }
+
+  void _init() {
+    _isValid = false;
+    _validationMessages.clear();
+
+    if (configured != null) {
+      _validationMessages.add('android-studio-dir = $configured');
+    }
+
+    if (!fs.isDirectorySync(directory)) {
+      _validationMessages.add('Android Studio not found at $directory');
+      return;
+    }
+
+    Version latestGradleVersion;
+
+    List<FileSystemEntity> gradlePaths;
+    try {
+      gradlePaths = fs.directory(fs.path.join(directory, 'gradle')).listSync();
+      for (FileSystemEntity entry in gradlePaths.where((FileSystemEntity e) =>
+          e.basename.startsWith('gradle-') && e is Directory)) {
+        Version version =
+            new Version.parse(entry.basename.substring('gradle-'.length));
+        if (latestGradleVersion == null || version > latestGradleVersion) {
+          latestGradleVersion = version;
+          if (version >= minGradleVersion) {
+            _gradlePath = entry.path;
+          }
+        }
+      }
+    } catch (e) {
+      printTrace('Unable to determine Gradle version: $e');
+    }
+
+    if (latestGradleVersion == null) {
+      _validationMessages.add('Gradle not found.');
+    } else if (_gradlePath == null) {
+      _validationMessages.add('Gradle version $minGradleVersion required. '
+          'Found version $latestGradleVersion.');
```

Comment:
nit: it's more readable if the parts of a multi-part string are vertically aligned.